File: hogtrace/ast.py
"""
HogTrace Abstract Syntax Tree (AST) data structures.

These classes represent the parsed structure of HogTrace programs
in a clean, easy-to-use format.
"""

from dataclasses import dataclass, field
from typing import Optional, Union
from enum import Enum


# ===== Expressions =====

class ExpressionType(Enum):
    """Types of expressions in HogTrace"""
    LITERAL = "literal"
    IDENTIFIER = "identifier"
    FIELD_ACCESS = "field_access"
    INDEX_ACCESS = "index_access"
    FUNCTION_CALL = "function_call"
    REQUEST_VAR = "request_var"
    BINARY_OP = "binary_op"
    UNARY_OP = "unary_op"


@dataclass
class Expression:
    """Base class for all expressions"""
    type: ExpressionType
    raw: str  # Raw text representation

    def __str__(self):
        return self.raw


@dataclass
class Literal(Expression):
    """Literal value (int, float, string, bool, None)"""
    value: Union[int, float, str, bool, None]

    def __init__(self, value: Union[int, float, str, bool, None], raw: str):
        super().__init__(ExpressionType.LITERAL, raw)
        self.value = value


@dataclass
class Identifier(Expression):
    """Variable identifier (args, arg0, retval, etc.)"""
    name: str

    def __init__(self, name: str):
        super().__init__(ExpressionType.IDENTIFIER, name)
        self.name = name


@dataclass
class FieldAccess(Expression):
    """Object field access (obj.field)"""
    object: Expression
    field: str

    def __init__(self, object: Expression, field: str, raw: str):
        super().__init__(ExpressionType.FIELD_ACCESS, raw)
        self.object = object
        self.field = field


@dataclass
class IndexAccess(Expression):
    """Array/dict index access (obj[index])"""
    object: Expression
    index: Expression

    def __init__(self, object: Expression, index: Expression, raw: str):
        super().__init__(ExpressionType.INDEX_ACCESS, raw)
        self.object = object
        self.index = index


@dataclass
class FunctionCall(Expression):
    """Function call (timestamp(), len(args))"""
    function: str
    arguments: list[Expression]

    def __init__(self, function: str, arguments: list[Expression], raw: str):
        super().__init__(ExpressionType.FUNCTION_CALL, raw)
        self.function = function
        self.arguments = arguments


@dataclass
class RequestVar(Expression):
    """Request-scoped variable ($req.var or $request.var)"""
    name: str
    prefix: str  # "req" or "request"

    def __init__(self, name: str, prefix: str, raw: str):
        super().__init__(ExpressionType.REQUEST_VAR, raw)
        self.name = name
        self.prefix = prefix


@dataclass
class BinaryOp(Expression):
    """Binary operation (a + b, a == b, etc.)"""
    operator: str
    left: Expression
    right: Expression

    def __init__(self, operator: str, left: Expression, right: Expression, raw: str):
        super().__init__(ExpressionType.BINARY_OP, raw)
        self.operator = operator
        self.left = left
        self.right = right


@dataclass
class UnaryOp(Expression):
    """Unary operation (!expr)"""
    operator: str
    operand: Expression

    def __init__(self, operator: str, operand: Expression, raw: str):
        super().__init__(ExpressionType.UNARY_OP, raw)
        self.operator = operator
        self.operand = operand


# ===== Actions =====

class ActionType(Enum):
    """Types of actions in HogTrace"""
    CAPTURE = "capture"
    ASSIGNMENT = "assignment"
    SAMPLE = "sample"


@dataclass
class Action:
    """Base class for all actions"""
    type: ActionType


@dataclass
class CaptureAction(Action):
    """Capture/send action"""
    function: str  # "capture" or "send"
    arguments: list[Expression]
    named_arguments: dict[str, Expression]

    def __init__(self, function: str, arguments: list[Expression] = None,
                 named_arguments: dict[str, Expression] = None):
        super().__init__(ActionType.CAPTURE)
        self.function = function
        self.arguments = arguments or []
        self.named_arguments = named_arguments or {}

    def __str__(self):
        args = []
        args.extend(str(arg) for arg in self.arguments)
        args.extend(f"{name}={expr}" for name, expr in self.named_arguments.items())
        return f"{self.function}({', '.join(args)})"


@dataclass
class AssignmentAction(Action):
    """Assignment to request-scoped variable"""
    variable: RequestVar
    value: Expression

    def __init__(self, variable: RequestVar, value: Expression):
        super().__init__(ActionType.ASSIGNMENT)
        self.variable = variable
        self.value = value

    def __str__(self):
        return f"{self.variable} = {self.value}"


@dataclass
class SampleAction(Action):
    """Sampling directive"""
    spec: str  # "10%", "1/100", etc.
    is_percentage: bool
    value: Optional[float] = None  # Parsed percentage (0.0-1.0)
    numerator: Optional[int] = None  # For ratio sampling
    denominator: Optional[int] = None

    def __init__(self, spec: str, is_percentage: bool, value: float = None,
                 numerator: int = None, denominator: int = None):
        super().__init__(ActionType.SAMPLE)
        self.spec = spec
        self.is_percentage = is_percentage
        self.value = value
        self.numerator = numerator
        self.denominator = denominator

    def __str__(self):
        return f"sample {self.spec}"


# ===== Probe Components =====

@dataclass
class ProbeSpec:
    """Probe specification (provider:module.function:probe_point)"""
    provider: str  # "fn", "py", etc.
    module_function: str  # "myapp.users.create_user"
    probe_point: str  # "entry", "exit", "entry+10", etc.
    full_spec: str  # Complete specification string

    def __str__(self):
        return self.full_spec


@dataclass
class Predicate:
    """Predicate (guard condition)"""
    expression: Expression

    def __str__(self):
        return f"/ {self.expression} /"


@dataclass
class Probe:
    """A single probe definition"""
    spec: ProbeSpec
    predicate: Optional[Predicate]
    actions: list[Action]

    def __init__(self, spec: ProbeSpec, predicate: Optional[Predicate] = None,
                 actions: list[Action] = None):
        self.spec = spec
        self.predicate = predicate
        self.actions = actions or []

    def __str__(self):
        result = [str(self.spec)]
        if self.predicate:
            result.append(str(self.predicate))
        result.append("{")
        for action in self.actions:
            result.append(f"    {action};")
        result.append("}")
        return "\n".join(result)


# ===== Program =====

@dataclass
class Program:
    """A complete HogTrace program (collection of probes)"""
    probes: list[Probe]

    def __init__(self, probes: Optional[list[Probe]] = None):
        self.probes = probes or []

    def __str__(self):
        return "\n\n".join(str(probe) for probe in self.probes)

    def __len__(self):
        return len(self.probes)

    def __iter__(self):
        return iter(self.probes)

    def __getitem__(self, index):
        return self.probes[index]

File: vm/src/parser/ast.rs
//! Abstract Syntax Tree for HogTrace language

use super::lexer::Span;
use std::fmt;

/// Top-level AST node representing a complete HogTrace program
#[derive(Debug, Clone)]
pub struct AstProgram {
    pub probes: Vec<AstProbe>,
    pub span: Span,
}

/// A probe definition
#[derive(Debug, Clone)]
pub struct AstProbe {
    pub spec: ProbeSpec,
    pub predicate: Option<AstExpr>,
    pub body: Vec<AstStatement>,
    pub span: Span,
}

/// Probe specification
#[derive(Debug, Clone)]
pub struct ProbeSpec {
    pub provider: Provider,
    pub module_function: ModuleFunction,
    pub probe_point: ProbePoint,
    pub span: Span,
}

/// Provider type (fn or py)
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Provider {
    Fn,
    Py,
}

/// Module and function path
#[derive(Debug, Clone)]
pub struct ModuleFunction {
    pub parts: Vec<ModulePart>,
    pub span: Span,
}

#[derive(Debug, Clone)]
pub enum ModulePart {
    Ident(String),
    Wildcard,
}

/// Probe point (entry, exit, etc.)
#[derive(Debug, Clone)]
pub enum ProbePoint {
    Entry,
    Exit,
    EntryOffset(i64),
    ExitOffset(i64),
}

/// Statements in action blocks
#[derive(Debug, Clone)]
pub enum AstStatement {
    Assignment {
        var: RequestVar,
        value: AstExpr,
        span: Span,
    },
    Sample {
        spec: SampleSpec,
        span: Span,
    },
    Capture {
        is_send: bool,
        args: CaptureArgs,
        span: Span,
    },
}

/// Request-scoped variable
#[derive(Debug, Clone)]
pub struct RequestVar {
    pub is_request: bool, // true for $request, false for $req
    pub field: String,
    pub span: Span,
}

/// Sample specification
#[derive(Debug, Clone)]
pub enum SampleSpec {
    Percentage(i64),
    Ratio { numerator: i64, denominator: i64 },
}

/// Capture arguments
#[derive(Debug, Clone)]
pub enum CaptureArgs {
    Positional(Vec<AstExpr>),
    Named(Vec<NamedArg>),
}

#[derive(Debug, Clone)]
pub struct NamedArg {
    pub name: String,
    pub value: AstExpr,
    pub span: Span,
}

/// Expressions
#[derive(Debug, Clone)]
pub enum AstExpr {
    // Literals
    Int {
        value: i64,
        span: Span,
    },
    Float {
        value: f64,
        span: Span,
    },
    String {
        value: String,
        span: Span,
    },
    Bool {
        value: bool,
        span: Span,
    },
    None {
        span: Span,
    },

    // Identifiers
    Ident {
        name: String,
        span: Span,
    },
    RequestVar {
        var: RequestVar,
        span: Span,
    },

    // Binary operations
    Binary {
        op: BinaryOp,
        left: Box<AstExpr>,
        right: Box<AstExpr>,
        span: Span,
    },

    // Unary operations
    Unary {
        op: UnaryOp,
        expr: Box<AstExpr>,
        span: Span,
    },

    // Field access (obj.field)
    FieldAccess {
        object: Box<AstExpr>,
        field: String,
        span: Span,
    },

    // Index access (obj[index])
    IndexAccess {
        object: Box<AstExpr>,
        index: Box<AstExpr>,
        span: Span,
    },

    // Function call
    Call {
        function: String,
        args: Vec<AstExpr>,
        span: Span,
    },
}

impl AstExpr {
    pub fn span(&self) -> Span {
        match self {
            AstExpr::Int { span, .. }
            | AstExpr::Float { span, .. }
            | AstExpr::String { span, .. }
            | AstExpr::Bool { span, .. }
            | AstExpr::None { span }
            | AstExpr::Ident { span, .. }
            | AstExpr::RequestVar { span, .. }
            | AstExpr::Binary { span, .. }
            | AstExpr::Unary { span, .. }
            | AstExpr::FieldAccess { span, .. }
            | AstExpr::IndexAccess { span, .. }
            | AstExpr::Call { span, .. } => *span,
        }
    }
}

/// Binary operators
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BinaryOp {
    // Arithmetic
    Add,
    Sub,
    Mul,
    Div,
    Mod,

    // Comparison
    Lt,
    Gt,
    LtEq,
    GtEq,
    Eq,
    NotEq,

    // Logical
    And,
    Or,
}

/// Unary operators
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum UnaryOp {
    Not,
}

// Display implementations for debugging and testing

impl fmt::Display for Provider {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Provider::Fn => write!(f, "fn"),
            Provider::Py => write!(f, "py"),
        }
    }
}

impl fmt::Display for ProbePoint {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ProbePoint::Entry => write!(f, "entry"),
            ProbePoint::Exit => write!(f, "exit"),
            ProbePoint::EntryOffset(n) => write!(f, "entry+{}", n),
            ProbePoint::ExitOffset(n) => write!(f, "exit+{}", n),
        }
    }
}

impl fmt::Display for ModuleFunction {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        for (i, part) in self.parts.iter().enumerate() {
            if i > 0 {
                write!(f, ".")?;
            }
            match part {
                ModulePart::Ident(s) => write!(f, "{}", s)?,
                ModulePart::Wildcard => write!(f, "*")?,
            }
        }
        Ok(())
    }
}

impl fmt::Display for BinaryOp {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            BinaryOp::Add => write!(f, "+"),
            BinaryOp::Sub => write!(f, "-"),
            BinaryOp::Mul => write!(f, "*"),
            BinaryOp::Div => write!(f, "/"),
            BinaryOp::Mod => write!(f, "%"),
            BinaryOp::Lt => write!(f, "<"),
            BinaryOp::Gt => write!(f, ">"),
            BinaryOp::LtEq => write!(f, "<="),
            BinaryOp::GtEq => write!(f, ">="),
            BinaryOp::Eq => write!(f, "=="),
            BinaryOp::NotEq => write!(f, "!="),
            BinaryOp::And => write!(f, "&&"),
            BinaryOp::Or => write!(f, "||"),
        }
    }
}

impl fmt::Display for UnaryOp {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            UnaryOp::Not => write!(f, "!"),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::lexer::Position;

    fn dummy_span() -> Span {
        Span::new(Position::start(), Position::start())
    }

    #[test]
    fn test_provider_display() {
        assert_eq!(Provider::Fn.to_string(), "fn");
        assert_eq!(Provider::Py.to_string(), "py");
    }

    #[test]
    fn test_probe_point_display() {
        assert_eq!(ProbePoint::Entry.to_string(), "entry");
        assert_eq!(ProbePoint::Exit.to_string(), "exit");
        assert_eq!(ProbePoint::EntryOffset(10).to_string(), "entry+10");
        assert_eq!(ProbePoint::ExitOffset(5).to_string(), "exit+5");
    }

    #[test]
    fn test_module_function_display() {
        let mf = ModuleFunction {
            parts: vec![
                ModulePart::Ident("myapp".to_string()),
                ModulePart::Ident("users".to_string()),
                ModulePart::Ident("create".to_string()),
            ],
            span: dummy_span(),
        };
        assert_eq!(mf.to_string(), "myapp.users.create");
    }

    #[test]
    fn test_module_function_with_wildcard() {
        let mf = ModuleFunction {
            parts: vec![
                ModulePart::Ident("myapp".to_string()),
                ModulePart::Wildcard,
            ],
            span: dummy_span(),
        };
        assert_eq!(mf.to_string(), "myapp.*");
    }

    #[test]
    fn test_binary_op_display() {
        assert_eq!(BinaryOp::Add.to_string(), "+");
        assert_eq!(BinaryOp::Eq.to_string(), "==");
        assert_eq!(BinaryOp::And.to_string(), "&&");
    }

    #[test]
    fn test_unary_op_display() {
        assert_eq!(UnaryOp::Not.to_string(), "!");
    }

    #[test]
    fn test_ast_expr_span() {
        let span = dummy_span();
        let expr = AstExpr::Int { value: 42, span };
        assert_eq!(expr.span(), span);
    }

    #[test]
    fn test_request_var_construction() {
        let var = RequestVar {
            is_request: false,
            field: "user_id".to_string(),
            span: dummy_span(),
        };
        assert!(!var.is_request);
        assert_eq!(var.field, "user_id");
    }

    #[test]
    fn test_sample_spec_variants() {
        let percentage = SampleSpec::Percentage(50);
        match percentage {
            SampleSpec::Percentage(n) => assert_eq!(n, 50),
            _ => panic!("Expected percentage"),
        }

        let ratio = SampleSpec::Ratio {
            numerator: 1,
            denominator: 10,
        };
        match ratio {
            SampleSpec::Ratio { numerator, denominator } => {
                assert_eq!(numerator, 1);
                assert_eq!(denominator, 10);
            }
            _ => panic!("Expected ratio"),
        }
    }

    #[test]
    fn test_capture_args_variants() {
        let positional = CaptureArgs::Positional(vec![]);
        match positional {
            CaptureArgs::Positional(_) => (),
            _ => panic!("Expected positional"),
        }

        let named = CaptureArgs::Named(vec![]);
        match named {
            CaptureArgs::Named(_) => (),
            _ => panic!("Expected named"),
        }
    }

    #[test]
    fn test_ast_statement_variants() {
        let assign = AstStatement::Assignment {
            var: RequestVar {
                is_request: false,
                field: "x".to_string(),
                span: dummy_span(),
            },
            value: AstExpr::Int {
                value: 42,
                span: dummy_span(),
            },
            span: dummy_span(),
        };
        match assign {
            AstStatement::Assignment { .. } => (),
            _ => panic!("Expected assignment"),
        }

        let sample = AstStatement::Sample {
            spec: SampleSpec::Percentage(10),
            span: dummy_span(),
        };
        match sample {
            AstStatement::Sample { .. } => (),
            _ => panic!("Expected sample"),
        }

        let capture = AstStatement::Capture {
            is_send: false,
            args: CaptureArgs::Positional(vec![]),
            span: dummy_span(),
        };
        match capture {
            AstStatement::Capture { .. } => (),
            _ => panic!("Expected capture"),
        }
    }

    #[test]
    fn test_binary_expr_construction() {
        let left = Box::new(AstExpr::Int {
            value: 1,
            span: dummy_span(),
        });
        let right = Box::new(AstExpr::Int {
            value: 2,
            span: dummy_span(),
        });

        let binary = AstExpr::Binary {
            op: BinaryOp::Add,
            left,
            right,
            span: dummy_span(),
        };

        match binary {
            AstExpr::Binary { op, .. } => assert_eq!(op, BinaryOp::Add),
            _ => panic!("Expected binary"),
        }
    }

    #[test]
    fn test_unary_expr_construction() {
        let expr = Box::new(AstExpr::Bool {
            value: true,
            span: dummy_span(),
        });

        let unary = AstExpr::Unary {
            op: UnaryOp::Not,
            expr,
            span: dummy_span(),
        };

        match unary {
            AstExpr::Unary { op, .. } => assert_eq!(op, UnaryOp::Not),
            _ => panic!("Expected unary"),
        }
    }

    #[test]
    fn test_field_access_construction() {
        let obj = Box::new(AstExpr::Ident {
            name: "user".to_string(),
            span: dummy_span(),
        });

        let access = AstExpr::FieldAccess {
            object: obj,
            field: "name".to_string(),
            span: dummy_span(),
        };

        match access {
            AstExpr::FieldAccess { field, .. } => assert_eq!(field, "name"),
            _ => panic!("Expected field access"),
        }
    }

    #[test]
    fn test_index_access_construction() {
        let obj = Box::new(AstExpr::Ident {
            name: "arr".to_string(),
            span: dummy_span(),
        });
        let index = Box::new(AstExpr::Int {
            value: 0,
            span: dummy_span(),
        });

        let access = AstExpr::IndexAccess {
            object: obj,
            index,
            span: dummy_span(),
        };

        match access {
            AstExpr::IndexAccess { .. } => (),
            _ => panic!("Expected index access"),
        }
    }

    #[test]
    fn test_function_call_construction() {
        let call = AstExpr::Call {
            function: "test".to_string(),
            args: vec![],
            span: dummy_span(),
        };

        match call {
            AstExpr::Call { function, args, .. } => {
                assert_eq!(function, "test");
                assert_eq!(args.len(), 0);
            }
            _ => panic!("Expected call"),
        }
    }
}
